### PR TITLE
Build tiller image for various architectures

### PR DESCRIPTION
**Pull request by alicefr**
Helm already supports cross-compiling and the parent image `alpine:3.7` is multi-arch. Therefore, building the `tiller` image for other architectures rather than x86_64 is straightforward. The docker images can be builti by using `make docker-build-images `

I also generalized the building process of a single docker image. The default target architecture remains `amd64`. However, another architecture could be set as the target. For example
```sh
$ make GO_ARCH=s390x docker-build
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -3,6 +3,7 @@ IMAGE_PREFIX      ?= kubernetes-helm
 SHORT_NAME        ?= tiller
 SHORT_NAME_RUDDER ?= rudder
 TARGETS           ?= darwin/amd64 linux/amd64 linux/386 linux/arm linux/arm64 linux/ppc64le windows/amd64
```
Comment (stefsuse) [inline Makefile]: I'm thinking it probably makes sense to merge your other PR with this change   #4190

**Review by stefsuse — COMMENTED** [review]
I saw this in helm-dev and just wanted to chime in though I am not a maintainer. Overall LGTM, thanks for doing this.

**Review by bacongobbler — COMMENTED** [review]
This will allow users to build cross-arch tiller images from source (which is great!), but would it be possible to hook this up to CI so we can publish these images to gcr.io?

Perhaps we should hold off on this until we've completed the [transfer over to the CNCF](https://www.linuxfoundation.org/blog/cncf-to-host-helm/) as they may want us to migrate our assets somewhere else. Any other maintainers have opinions on this?

**Comment by bacongobbler** [comment]
For reference, this is how we build and push the release assets today on a merge to master: https://github.com/kubernetes/helm/blob/4e5139b67c53bc87271e456cbd658c9f448f19d7/.circleci/deploy.sh#L53-L64

**Comment by stefsuse** [comment]
Sure. Update deploy.sh to accept an arg for arch. Then update .circleci.yaml to pass the $arch string at the deploy step (not sure how you do that with circle specifically though). On its own this change won't change the current process but it at least allows for the option. If CNCF decides to host the builds elsewhere they just update deploy.sh? Just my two cents.

File: Makefile (hunk 1 above)
Comment (alicefr) [reply Makefile]: I guess this modification is a little bit more complex. In https://github.com/kubernetes/helm/pull/4190 I simply added an architecture

**Comment by alicefr** [comment]
@bacongobbler I didn't think to add into circleci. But that also depends how now you want to call your images. With kubernetes style you could append the architecture at the end of the image name or another possibility could be to create a multi-arch image. The advantage of the second approach would be that you do not have to change the name of the tiller image. Would you pack everything in a PR or better a second one?

I'm traveling this week and then I will take some vacation. Therefore, the discussion will be a bit slow down.

**Comment by bacongobbler** [comment]
After doing a little research, there appears to be some new work available where we can distribute arch-specific images ([talks at DockerCon 2017](https://www.youtube.com/watch?v=nrBYUw1Pz5I&t=20m)) and then create a manifest list that points to these individual images. That way, `docker run gcr.io/kubernetes-helm/tiller` will pull the right architecture required for the OS running the image. That looks like a good approach in the long run to me! :)

The first part would be the easiest to tackle: we start by releasing arch-specific images of tiller like `gcr.io/kubernetes-helm/tiller-arm`. The second part would be tricky, as we would basically need to replace `gcr.io/kubernetes-helm/tiller` with a manifest list, and that *might* break stuff if it's still an experimental feature.

I think taking the approach where we distribute images with the architecture being appended to the end of the image name would be a good first step, and then once the latter features become more stable we can start adding support for that. How do you feel about that approach?

**Comment by alicefr** [comment]
yes it sounds reasonable. I haven't tried the `docker manifest` command yet. I know that the Docker developers used the [manifest-tool](https://github.com/estesp/manifest-tool) to create the multi-arch image. But I understand that you want to wait for the manifest command.

Do you want that I include the modifications at `helm/.circleci/deploy.sh` in this PR?

**Comment by alicefr** [comment]
hi, I just realized that the PR is actually wrong, sorry. It actually cross-compiles the binaries for the different images, but the parent image remains `x86`

**Pull request closed without merging**